creature: fix dead water creature placement

lahm86 opened the pull request:
Resolves #1031.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This moves the logic for shifting dead rats and crocodiles into shared functions in `creature.c` to resolve inconsistencies between the two and to cover all scenarios. Once a dead hybrid creature has been converted to its land type, it will remain in that state permanently.

I've also introduced an idiom in `Item_SwitchToObjAnim` that if you pass a negative frame number, this will be inferred as meaning the frame should be relative from the end of the animation.

lahm86 committed:
creature: fix dead water creature placement

Moves the logic for shifting dead water creatures into shared functions
to resolve inconsistencies and to cover all scenarios. Once a dead
hybrid creature has been converted to its land type, it will remain in
that state permanently.

Resolves #1031.

rr- commented on src/game/objects/creatures/rat.c:
```diff
@@ -49,6 +49,13 @@ typedef enum {
+                                       .land.death_anim = RAT_DIE_ANIM,
+                                       .land.death_state = RAT_DEATH,
+                                       .water.id = O_VOLE,
+                                       .water.active_anim = VOLE_EMPTY };
```
Will it be formatted any better if we put a trailing comma inside the struct?

lahm86 replied:
It did, thanks. I've done the same in `crocodile.c` for consistency.

rr- commented on src/game/creature.c:
```diff
@@ -706,3 +711,80 @@ bool Creature_ShootAtLara(
+    return item->object_number == info->land.id
+        ? Creature_TestLandState(item_num, wh, info)
+        : Creature_TestWaterState(item_num, wh, info);
+}
```
The name `Test` suggests that the function doesn't mutate the global state, while it changes up item animations and object numbers. Furthermore `Hybrid` is unclear until the reader already knows the concept behind it. How about `Creature_EnsureHabitat` or similar?

lahm86 replied:
Good point, thanks. Now renamed as above, and the statics have been renamed similarly too. I've also refactored slightly to call `Room_GetWaterHeight` in the wrapper function rather than in each state function for a little less repetition.

lahm86 committed:
creature: rename habitat functions

Renames the functions that manipulate creature state for land/water,
plus some additional reformatting.

Part of #1031.

walkawayy commented on src/game/objects/creatures/rat.c:
```diff
@@ -49,6 +49,15 @@ typedef enum {
+    .land.death_state = RAT_DEATH,
+    .water.id = O_VOLE,
+    .water.active_anim = VOLE_EMPTY,
+};
```
Why does the water state not need a death anim/state?

lahm86 replied:
We never switch to the water death state in the environment checks, so it's not needed. These are the only possible scenarios:
- Killed in water (normal control uses water death anim)
  - Room drained - switch to land death
  - Room re-flooded - remain in land death
- Killed on land
  - Room flooded - remain in land death
  - Room re-drained - remain in land death

The only difference is for crocodiles - if they're killed in water and that room is then drained, but there is still another water room below, they'll remain in the water anim death state.

walkawayy reviewed (approved):
All testing LGTM.
